### PR TITLE
Handle more cases with the new entry point pattern

### PR DESCRIPTION
**Pull request by davidfowl**
- Handle an exception being thrown from main before start is called and make sure it propagates to the WebApplicationFactory.
- Don't hang if the application doesn't call Start before it completes.

### CODE REVIEW

**Review by javiercn — APPROVED** [review]

**Comment by davidfowl** [comment]
/azp run

I just fixed this https://github.com/dotnet/runtime/pull/54133 and this hang might be related.

File: AspNetCore.sln [hunk 1]
```diff
@@ -1624,6 +1624,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "PhotinoTestApp", "src\Compo
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.AspNetCore.Components.WebView.Photino", "src\Components\WebView\Samples\PhotinoPlatform\src\Microsoft.AspNetCore.Components.WebView.Photino.csproj", "{B1AA24A4-5E02-4DC1-B57F-6EB03F91E4DD}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "SimpleWebSiteWithWebApplicationBuilderException", "src\Mvc\test\WebSites\SimpleWebSiteWithWebApplicationBuilderException\SimpleWebSiteWithWebApplicationBuilderException.csproj", "{5C641396-7E92-4F5C-A5A1-B4CDF480539B}"
```
Comment (captainsafia) [inline AspNetCore.sln:1627]: Do we need to create a new test project for each of these scenarios or is there a way for us to have one test project with multiple apps for each scenario (app runs fine, app throws exception, etc)?

Comment (davidfowl) [reply AspNetCore.sln:1627]: We could have a single app that changed behavior based on arguments.

**Comment by davidfowl** [comment]
This is still hanging, I need to look at the dump

**Comment by davidfowl** [comment]
I just realized we're still on an older runtime build that doesn't have my latest fixes.

@dougbu it turns out [this](https://github.com/dotnet/aspnetcore/pull/33524) change still carries and older version of HostFactoryResolver. We're still pretty behind even though the PR was 11 hours ago, the content is 2 days old.

**Comment by dougbu** [comment]
> I just realized we're still on an older runtime build that doesn't have my latest fixes.

This relates to dotnet/arcade#7532 but @epananth at least temporarily worked around that and got new dotnet/runtime versions into the '.NET 6' and '.NET 6 Preview 6' channels. You're waiting for #33560 in main and (for #33574) #33571 in release/6.0-preview6. Both are blocked by our need to react to dotnet/runtime#54147.

File: src/Mvc/Mvc.Testing/src/DeferredHostBuilder.cs [hunk 2]
```diff
@@ -109,20 +128,18 @@ public ValueTask DisposeAsync()
...
-                using var reg = cancellationToken.UnsafeRegister(_ => tcs.TrySetCanceled(), null);
+                using var reg = cancellationToken.UnsafeRegister(_ => _hostStartedTcs.TrySetCanceled(), null);
 
                 // REVIEW: This will deadlock if the application creates the host but never calls start. This is mitigated by the cancellationToken
                 // but it's rarely a valid token for Start
-                _host.Services.GetRequiredService<IHostApplicationLifetime>().ApplicationStarted.UnsafeRegister(_ => tcs.TrySetResult(), null);
+                using var reg2 = _host.Services.GetRequiredService<IHostApplicationLifetime>().ApplicationStarted.UnsafeRegister(_ => _hostStartedTcs.TrySetResult(), null);
 
-                return tcs.Task;
+                await _hostStartedTcs.Task;
```
Comment (halter73) [inline src/Mvc/Mvc.Testing/src/DeferredHostBuilder.cs:142]: Out of curiosity, why did you change this from `return` to `await`?

Comment (halter73) [reply src/Mvc/Mvc.Testing/src/DeferredHostBuilder.cs:142]: The using. Doh!
